### ISSUE TITLE
Non-Virtual Setup Analyzer gives false positive when complex argument constraint used in indexer

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1072]
Found in [the Nancy codebase](https://github.com/NancyFx/Nancy/blob/be4f8d42076e4e568a3742715437868e6c7d05af/test/Nancy.ViewEngines.Razor.Tests/TextResourceFinderFixture.cs#L68).
Can be reproduced by

```c#
namespace TheNamespace
{
    using FakeItEasy;

    public interface IHaveAnIndexer { string this[int context] { get; } }

    public class TheClass
    {
        public void Test()
        {
            var fake = A.Fake<IHaveAnIndexer>();
            A.CallTo(() => fake[A<int>._]).MustHaveHappened();
        }
    }
}
```

**Comment by blairconrad** [comment]
Does not occur when `A<int>._` is replaced by a literal, or when the indexer is replaced by a method.

**Comment by blairconrad** [comment]
Seems to me that we're getting the false positive because we're looking for ancestors of `A<int>._`, looking for the first `InvocationExpressionSyntax`, which is `A.CallTo(() => fake[A<int>._])`, but there's an intervening `ElementAccessExpressionSyntax` node (`fake[A<int>._]`) that should have stopped us walking up the tree.
In VB.NET, the equivalent node is an `InvocationExpressionSyntax` (`fake(A(Of Integer).Ignored)`.)

**Comment by thomaslevesque** [comment]
Yes, so the issue should only happen for C#, not VB.
I can fix it when I get back, so the fix can be included in 3.3.0 (unless you want to do it of course)

**Comment by blairconrad** [comment]
You are probably all vacationy so didn't notice that I self-assigned and marked in-progress. I am on top of this!
I enjoy an opportunity to demonstrate my complete lack of understanding of analyzer issues, submit a fix, and have you tell me it's wrong!
Srsly, though, I'll upload a commit in a few minutes.

**Comment by blairconrad** [comment]
This issue was fixed in [Release 3.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/3.3.0).
